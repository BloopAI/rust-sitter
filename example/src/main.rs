use std::io::Write;

#[rust_sitter::grammar]
pub mod arithmetic_grammar {
    #[rust_sitter::language]
    #[derive(Debug)]
    pub enum Expression {
        Number(
            #[rust_sitter::leaf(pattern = r"\d+", transform = |v: &str| v.parse().unwrap())] i32,
        ),
        #[rust_sitter::prec_left(1)]
        Sub(
            Box<Expression>,
            #[rust_sitter::leaf(text = "-", transform = |_v| ())] (),
            Box<Expression>,
        ),
        #[rust_sitter::prec_left(2)]
        Mul(
            Box<Expression>,
            #[rust_sitter::leaf(text = "*", transform = |_v| ())] (),
            Box<Expression>,
        ),
    }

    #[rust_sitter::extra]
    struct Whitespace {
        #[rust_sitter::leaf(pattern = r"\s", transform = |_v| ())]
        _whitespace: (),
    }
}

fn main() {
    let stdin = std::io::stdin();

    loop {
        print!("> ");
        std::io::stdout().flush().unwrap();

        let mut input = String::new();
        stdin.read_line(&mut input).unwrap();
        let input = input.trim();
        if input.is_empty() {
            break;
        }

        println!("{:?}", arithmetic_grammar::parse(input));
    }
}

#[cfg(test)]
mod tests {
    use super::arithmetic_grammar;

    #[test]
    fn arithmetic_grammar() {
        insta::assert_debug_snapshot!(arithmetic_grammar::parse("1"));
        insta::assert_debug_snapshot!(arithmetic_grammar::parse("1 - 2"));
        insta::assert_debug_snapshot!(arithmetic_grammar::parse("1 - 2 - 3"));
        insta::assert_debug_snapshot!(arithmetic_grammar::parse("1 - 2 * 3"));
        insta::assert_debug_snapshot!(arithmetic_grammar::parse("1 * 2 * 3"));
        insta::assert_debug_snapshot!(arithmetic_grammar::parse("1 * 2 - 3"));
    }
}
