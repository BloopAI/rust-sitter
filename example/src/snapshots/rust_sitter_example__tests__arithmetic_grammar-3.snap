---
source: example/src/main.rs
expression: "arithmetic_grammar::parse(\"1-2-3\")"
---
Sub(
    Sub(
        Number(
            1,
        ),
        (),
        Number(
            2,
        ),
    ),
    (),
    Number(
        3,
    ),
)
